### ISSUE TITLE
Test on OCID and OSD datasets

### ISSUE THREAD

**Issue by yellow07200** [chrisdxie/uois#13]
Hi, 
 I am trying to test my pretrained network on OCID and OSD datasets, can you provide the source file of preprocessing datasets to npy format?  Thanks!

**Comment by zl9501** [comment]
Same problem. @chrisdxie Could you please provide the testing script that can reproduce results as in your paper with the pre-trained model? Thanks.

**Comment by chrisdxie** [comment]
Hi, I have moved institutions so I no longer have access to the testing script. However, it should be pretty simple to re-implement. If you have all of your predictions and the ground truth, you can run the [multilabel_metrics()](https://github.com/chrisdxie/uois/blob/master/src/evaluation.py#L48) function and compute the mean over the dataset.

There is no need to preprocess the files to npy format. You can read the raw PCD files with something like [Open3D](http://www.open3d.org/) and load them in that way.

**Comment by yellow07200** [comment]
> 

Hi, @chrisdxie 
Thanks for your reply. I notice that xyz_img is required as inputs for training and testing, can you please guide how to convert pcd to xyz_img file? 

BTW, I see in data_loader.py, the focal length is required. If you also obtain xyz_img in this way, is it possible to provide the focal length you used? Many thanks!

**Comment by Qaoud** [comment]
Hello @chrisdxie,

I have the same problem, since I am trying to reproduce the paper's Tables to ensure that my setup works correctly.

I have extracted the predictions of the provided UOIS-Net-3D model, from the OSD (each of the 111 images) and OCID datasets (each of the 2346 images). To pre-process the images, I have adapted your code from the data_loder() and the uois_3D_example.ipynb notebook (I can also provide the code if needed).

Now I would like to understand how to compute the six metrics, and to which Table in the paper it is best to compare.

What I am doing now is to use the multilabel_metrics() function on each image, and then average each metric over the full dataset. However, in this way I get the following metrics (for OSD dataset):


<img width="857" alt="Screenshot 2022-01-10 at 10 48 27" src="https://user-images.githubusercontent.com/23742278/148774884-61b574c9-7156-4736-bc99-d1715d89d006.png">


which are different from the ones reported in the last row of Table II:


<img width="839" alt="Screenshot 2022-01-10 at 14 34 54" src="https://user-images.githubusercontent.com/23742278/148774850-1f5655f4-00d4-40e1-a2fd-1b89af7d318b.png">


Hence, I would like to ask if you can spot some difference in what I'm doing from what you did in the paper.

Thanks for your feedback!

**Comment by chrisdxie** [comment]
@yellow07200, PCD files contain the xyz images in them. There is no need for conversion. You can read them using something like Open3D. I only used focal length to backproject a depth image into an XYZ image.

@Qaoud, since I used a weird orientation for the XYZ images (left-handed coordinate system, where x points right, y points up, and z points forward), I typically flip the Y channel. If you load the OCID  You may need to multiple the y-channel by -1 to fix this. Additionally, any depth points that are NaN should be set to 0. Hopefully that helps.

**Comment by zl9501** [comment]
Hi @chrisdxie, I've done the same thing as @Qaoud and got almost the same results as him! I have carefully implemented the same normalization as in your code, set NaN depth points to 0, and multiplied the y-channel by -1, but still got 60 Boundary F-measure and 70 Objects F-measure for the OSD dataset. Could you possibly provide some suggestions to reproduce results as in the paper, thank you very much!

**Comment by chrisdxie** [comment]
@zl9501 Please check that the outputs of the XYZ image extracted from the OSD/OCID datasets look reasonable. For example, since OSD/OCID provide .pcd files which include XYZ images, there is no need to run the `util_.compute_xyz()` method. Some screenshots could help identify the issue.

**Comment by zl9501** [comment]
@chrisdxie Thanks very much for your kind reply! I used your visualization code to view the data I extract, all is ok. So I think maybe something goes wrong in the preprocess, below is my `__getitem__` function for the OSD dataset, if you see any problem, it will be very helpful to me. Thanks again.

![d5acfcb00a0745ff5e0b5af44dc1f9d](https://user-images.githubusercontent.com/68844981/149889962-69ad2da6-9148-4dba-a7c2-2228c4586d58.png)
![7840c4c79ec9d82a32181db7056aab6](https://user-images.githubusercontent.com/68844981/149890005-1070866e-ba5b-487f-836e-b66057060932.png)
![dfc0fc9961ab6c2b5548442bdf497e3](https://user-images.githubusercontent.com/68844981/149890039-fa3184df-b588-4c6e-b676-fb70dea58688.png)


> @zl9501 Please check that the outputs of the XYZ image extracted from the OSD/OCID datasets look reasonable. For example, since OSD/OCID provide .pcd files which include XYZ images, there is no need to run the `util_.compute_xyz()` method. Some screenshots could help identify the issue.

**Comment by chrisdxie** [comment]
Hi @zl9501, thanks for showing your code. It looks almost identical to mine, except that you use `pcl.load()` while I used the Open3D library to load a .pcd file. I don't imagine there is too much difference there.

Would you mind showing some outputs? Perhaps some intermediate outputs as well (e.g. the masks before applying the RRN)? Also, if you could supply the parameters you used, that also may be helpful.

**Comment by zl9501** [comment]
Hi @chrisdxie, thanks for your advice, I've tried to use the RGB image, and as you anticipated there wasn't too much difference.
Below shows the parameters I used (as is those in uois_3D_example.ipynb) and some intermediate outputs, thanks again for your kind help!

### 1. Parameters
![54149bb64421394fb61c04d2af2850c](https://user-images.githubusercontent.com/68844981/150066898-b49caf4d-d9d0-466e-a59c-5872272cf873.png)

### 2. Some intermediate outputs
#### Visualizations of images 1~4: 
![ab33d89ac1c69f18261516471c7114f](https://user-images.githubusercontent.com/68844981/150066969-63952e03-bd33-4583-88b0-f04d57bea6f8.png)
#### Metrics of images 1~4: 
![3cc5a043d5e78267d8695985efc15fc](https://user-images.githubusercontent.com/68844981/150067069-e855253d-ae29-45cc-839a-c0223038a8f1.png)

#### Visualizations of images 9~12:
![1a4d9c19b4e2eadfbce5e5cf2abe481](https://user-images.githubusercontent.com/68844981/150067095-50d5b061-c106-4287-911b-2d8c03a769fd.png)
#### Metrics of images 9~12: 
![83b22c092159ef46bef33a4ea6ad30d](https://user-images.githubusercontent.com/68844981/150067131-f531af36-6d3d-47e3-abef-f4e4972c5cd7.png)

**Comment by chrisdxie** [comment]
Hi @zl9501, thanks for providing some screenshots. I believe I have figured out the problem.

The evaluation function assumes that object indices start at 2, and the script will ignore classes that are 0/1 (background/table). OSD does not contain table labels, and object instances start at 1. OCID is different: 2 is the table, and the objects start at 3 (if I remember correctly). So, please load in the OSD/OCID labels with the corrected labels, and 

You can see this, for example in Image 10 where `obj_gt` = 1, when clearly there are 2 GT objects.

**Comment by zl9501** [comment]
Thank you @chrisdxie, I've followed your instructions and got similar results as in the paper. Thank you very much for your helpful suggestions and the time you have invested in this issue. Happy new year and best wishes to you!

**Comment by evinpinar** [comment]
Hey, were you able to replicate the results for OCID as well? I have done for OSD and it is fine, very similar to Table 2 on the paper. However, OCID is still different. I've been trying different starting indices, multiplying y channel with -1 and so.

Another interesting thing is, OCID has 2390 images instead of 2346 (downloading the online data https://www.acin.tuwien.ac.at/en/vision-for-robotics/software-tools/object-clutter-indoor-dataset/ and even there authors mentioned 2346 images but when I count the images in the folders, I end up with 2390 separate pcd/rgb/label triplets.) Do you recall anything on that aspect? 

Thanks!

**Comment by chrisdxie** [comment]
Hi @evinpinar, yes the results were replicated by zl9501 for OCID.

As I recall, OCID has 2390 images, not 2346.

**Comment by evinpinar** [comment]
Indeed, it seems there is a typo in one of the folders, instead of "pcd", folder name was "pd", resulting in reading 2346 images if you try to read from them.
